File: 242.有效的字母异位词.java
/*
 * @lc app=leetcode.cn id=242 lang=java
 *
 * [242] 有效的字母异位词
 */

// @lc code=start
import java.util.HashMap;


/** updated version */
// 优化1：仅使用一个map 实现算法，即统计s字符频数时 +1 ，统计t字符频数时-1；
// 优化2：先比较两个字符串长度，不一致直接返回false;
// 优化3：dic.getOrDefault(s.charAt(i), 0) 初始化字典   Returns the value to which the specified key is mapped, or defaultValue if this map contains no mapping for the key.



class Solution_1{
    public boolean isAnagram(String s, String t){
        if( s.length() != t.length()){
            return false;
        }
        HashMap<Character, Integer> dic = new HashMap<>();
        char[] s_c = s.toCharArray(), t_c = t.toCharArray();
        for(char c : s_c){
            dic.put(c, dic.getOrDefault(c, 0) + 1);
        }
        for(char c : t_c){
            dic.put(c, dic.getOrDefault(c, 0) - 1);
        }
        for(int i : dic.values()){
            if ( i != 0 ) return false;
        }

        return true;

    }
}




/** initial version */


class Solution {
    public boolean isAnagram(String s, String t) {
        HashMap<Character, Integer> s_h = new HashMap<>();
        HashMap<Character, Integer> t_h = new HashMap<>();
        char[] s_c = s.toCharArray();
        char[] t_c = t.toCharArray();
        for(char c : s_c ){
            if(s_h.get(c) == null){
                s_h.put(c, 0);
            }
            s_h.put(c, s_h.get(c) + 1);
        }

        for(char c : t_c){
            if(t_h.get(c) == null){
                t_h.put(c, 0);
            }
            t_h.put(c, t_h.get(c) + 1);
        }

        return s_h.equals(t_h);
    }
}
// @lc code=end

File: 242.有效的字母异位词.py
#
# @lc app=leetcode.cn id=242 lang=python3
#
# [242] 有效的字母异位词
#

# @lc code=start
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # 可以先判断一下长度是否相等，不相等则一定不是：
        if len(s) != len(t): return False
        # 字符串转List
        n = len(s)
        dic = {}
        for i in range(n):
            # hashmap用法:
            #判断是否存在:in
            if s[i] in dic:
                dic[s[i]] += 1
            else:
                dic[s[i]] = 1
        m = len(t)
        for j in range(m):
            if t[j] not in dic:
                return False
            else:
                dic[t[j]] -= 1
        for value in dic.values():
            if value != 0:
                return False
        return True

# @lc code=end

